### PR TITLE
Implement 'object ready' pattern in KeyVaultProvider and CosmosDBProvider

### LINKED ISSUE #93: Use member variable for Cosmos DB and Key Vault
## Description:
- Currently, the resource providers use an async initialize function that is awaited on 
- Suggestion to use a different pattern where there is a member variable that returns a promise indicating if the resource connection is ready.  Then, whenever the resource is needed, you can check by awaiting the ready property.
- References #11

### PR DESCRIPTION
**Pull request by codefoster**
## Type of PR

- [x] Documentation changes
  I did a search to determine that neither CosmosDBProvider nor KeyVaultProvider are mentioned in any markdown files, so there's no documentation on the instantiation or calling of these classes yet.

- [x] Code changes
- [x] Test changes
  No new functionality introduced - just a different calling pattern. Will test these in a separate task.

- [ ] CI-CD changes
  Did not explicitly test docker build because it takes so long. Code compiles and lints.

- [ ] GitHub Template changes

## Issues Closed or Referenced

- Closes #93

### CODE REVIEW

File: src/secrets/KeyVaultProvider.ts [hunk 1]
```diff
@@ -45,17 +47,14 @@ export class KeyVaultProvider {
...
-    private async _initialize() {
+    private async initialize() {
 
         // Use specified authentication type (either MSI or CLI)
         const creds: any = this.authType === "MSI" ?
             await msRestNodeAuth.loginWithAppServiceMSI({ resource: "https://vault.azure.net" }) :
             await msRestNodeAuth.AzureCliCredentials.create({ resource: "https://vault.azure.net" });
 
         this.client = new SecretClient(this.url, creds);
+        return;
```
Comment (bartr) [inline src/secrets/KeyVaultProvider.ts]: the client isn't fully initialized until you do a round trip to the key vault. With the fail early approach, you should test that here, throw a friendly error and exit(-1)

This is also a good place to put the retry logic necessary for pod identity.

Comment (aflinchb) [reply src/secrets/KeyVaultProvider.ts]: This is related to the conversation we had about where/when a failed credential authentication error would appear with this object ready pattern.  Maybe also/instead in config.ts when we await keyvault.ready -- if there is a way to catch/react to an authentication failure.

Comment (codefoster) [reply src/secrets/KeyVaultProvider.ts]: So I think we just add a `try` in the initializer so we can _explicitly_ log a friendly error, right?

Comment (codefoster) [reply src/secrets/KeyVaultProvider.ts]: I just pushed a commit adding a `try` around the keyvault initialize function.

Comment (bartr) [reply src/secrets/KeyVaultProvider.ts]: There's an issue for making this work with pod identity, so we can address the full round-trip then.

**Pull request merged**